### PR TITLE
Add a CI job to run tests with CRuby's test-all

### PR DESCRIPTION
**Pull request by k0kubun**
We've sometimes seen Prism sync to ruby/ruby fails CRuby's CI while it's passing Prism's CI. It'd be nice to catch such failures before merging a pull request.

This PR adds `.github/workflows/ruby-core.yml`. I copied this from `ruby/irb` (which was copied from `rubygems/rubygems`), added small improvements on it, and made the YAML closer to other ones in this repository.

### CODE REVIEW

**Comment by k0kubun** [comment]
The CI is failing, but that's because it's an actual problem that motivated this change. I believe Jemma is working on fixing it.

**Review by kddnewton — APPROVED** [review]

File: .github/workflows/ruby-core.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    branches:
+      - main
+  pull_request:
+
+jobs:
+  test-all:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Set up latest ruby head
+        uses: ruby/setup-ruby@250fcd6a742febb1123a77a841497ccaa8b9e939 # v1.152.0
```
Comment (eregon) [inline .github/workflows/ruby-core.yml:18]:
```suggestion
        uses: ruby/setup-ruby@v1
```

File: .github/workflows/ruby-core.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        with:
+          ruby-version: head
+          bundler: none
+      - name: Save latest buildable revision to environment
+        run: echo "REF=$(ruby -e 'print RUBY_REVISION')" >> $GITHUB_ENV
+      - uses: actions/checkout@8ade135a41bc03ea155e62e844d188df1ea18608 # v3.1.0
```
Comment (eregon) [inline .github/workflows/ruby-core.yml:24]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/ruby-core.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+          set -x
+          sudo apt-get update -q || :
+          sudo apt-get install --no-install-recommends -q -y build-essential libssl-dev libyaml-dev libreadline6-dev zlib1g-dev libncurses5-dev libffi-dev libgdbm-dev autoconf ruby
+      - name: Build Ruby
+        run: |
+          autoconf
+          ./configure -C --disable-install-doc
+          make -j2
+        working-directory: ruby/ruby
+      - uses: actions/checkout@8ade135a41bc03ea155e62e844d188df1ea18608 # v3.1.0
```
Comment (eregon) [inline .github/workflows/ruby-core.yml:45]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/ruby-core.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        run: echo "REF=$(ruby -e 'print RUBY_REVISION')" >> $GITHUB_ENV
+      - uses: actions/checkout@8ade135a41bc03ea155e62e844d188df1ea18608 # v3.1.0
+        with:
+          repository: ruby/ruby
+          path: ruby/ruby
+          fetch-depth: 1
+      - name: Checkout the latest buildable revision
+        run: |
+          git fetch --depth=1 origin ${{ env.REF }}
+          git switch -c ${{ env.REF }}
```
Comment (eregon) [inline .github/workflows/ruby-core.yml:32]: What does this do? Is it necessary? Isn't it enough to pick latest ruby/ruby master?

**Review by eregon — COMMENTED** [review]
Nice, I made a wrote comments to simplify/improve readability

File: .github/workflows/ruby-core.yml (hunk 1 above)
Comment (k0kubun) [reply .github/workflows/ruby-core.yml]: https://github.com/ruby/prism/pull/1697/commits/8180bbc85dd3ed3ee98ad50e946e5406128d464f

File: .github/workflows/ruby-core.yml (hunk 3 above)
Comment (k0kubun) [reply .github/workflows/ruby-core.yml]: https://github.com/ruby/prism/pull/1697/commits/8180bbc85dd3ed3ee98ad50e946e5406128d464f

File: .github/workflows/ruby-core.yml (hunk 2 above)
Comment (k0kubun) [reply .github/workflows/ruby-core.yml]: https://github.com/ruby/prism/pull/1697/commits/8180bbc85dd3ed3ee98ad50e946e5406128d464f

**Pull request merged**